pkg/cache/v2/cache.go

```diff
@@ -37,14 +37,14 @@ type ConfigWatcher interface {
 	// CreateWatch returns a new open watch from a non-empty request.
 	// An individual consumer normally issues a single open watch by each type URL.
 	//
-	// Value channel produces requested resources, once they are available.  If
-	// the channel is closed prior to cancellation of the watch, an unrecoverable
-	// error has occurred in the producer, and the consumer should close the
+	// Once the requested resources are available, they should be pushed into responses channel.
+	// If there are any failures - the error should be returned and the consumer should close the
 	// corresponding stream.
+	// Cache implementation should not close responses channel.
 	//
 	// Cancel is an optional function to release resources in the producer. If
 	// provided, the consumer may call this function multiple times.
-	CreateWatch(*Request) (value chan Response, cancel func())
+	CreateWatch(request *Request, responses chan<- Response) (cancel func(), err error)
```

Comment:
can we name the param to singular `responseChan`

Reply:
Will do later. Given the things you've said above - the best thing to do now is to agree if the approach I propose can be accepted at all. Cause mimic-ing exactly the same behavior with different implementation can be very tricky and if you depend on existing behavior - I really wonder if we can proceed.
Cause my changes are the foundation for more consistent ADS experience in subsequent PRs and don't necessarily improve non-ADS.

Reply:
Cause with previous implementation - there was a guarantee that if envoy does the subsequent request - the previous one will be cancelled and there would be no attempts to read from previous channel - only the new one will be used. With my PR that is changing - you might get 2 responses in a row if cancel() is called after something is already written to the channel.

Reply:
I dont have major concerns with the approach, since we dont use ADS and im not sure of the different nuances. Is it possible to demonstrate the scenario you're changing with a test case?

Reply:
Not within this PR, but in the end I want to implement this: 
https://www.envoyproxy.io/docs/envoy/latest/api-docs/xds_protocol#eventual-consistency-considerations, ie: reorder responses to make clusters go first, then endpoints, etc...
This can be done in a hacky way from server side, but it is reasonable to do on a cache side (when you do SetSnapshot and pass whole set of resources - at that point the ordering should be applied)

So in the end I want this to pass:

```go
func TestResponseOrdering(t *testing.T) {
       c := cache.NewSnapshotCache(true, group{}, logger{t: t})
       ch := make(chan cache.Response, 4)
       c.CreateWatch(&discovery.DiscoveryRequest{TypeUrl: rsrc.EndpointType, ResourceNames: names[rsrc.EndpointType]}, ch)
       c.CreateWatch(&discovery.DiscoveryRequest{TypeUrl: rsrc.ClusterType, ResourceNames: names[rsrc.ClusterType]}, ch)
       c.CreateWatch(&discovery.DiscoveryRequest{TypeUrl: rsrc.RouteType, ResourceNames: names[rsrc.RouteType]}, ch)
       c.CreateWatch(&discovery.DiscoveryRequest{TypeUrl: rsrc.ListenerType, ResourceNames: names[rsrc.ListenerType]}, ch)
       if err := c.SetSnapshot(key, snapshot); err != nil {
               t.Fatal(err)
       }
       expectedTypeUrls := []string{rsrc.ClusterType, rsrc.EndpointType, rsrc.ListenerType, rsrc.RouteType}
       var typeUrls []string
       for range expectedTypeUrls {
               resp := <-ch
               typeUrls = append(typeUrls, resp.GetRequest().TypeUrl)
       }
       if !reflect.DeepEqual(typeUrls, expectedTypeUrls) {
               t.Fatalf("incorrect types order: got %v, expected %v", typeUrls, []string{rsrc.ClusterType, rsrc.EndpointType, rsrc.ListenerType, rsrc.RouteType})
       }
}
```

Reply:
Single channel would allow to guarantee that it pass in ADS mode and will make it much more likely to pass in non-ADS mode (but without any guarantees)